rust/src/accountant.rs

```diff
@@ -0,0 +1,179 @@
+use crate::accumulator::{UsageAccumulator, UsageUnit};
+use crate::producer::{ClientError, KafkaConfig, KafkaProducer, Producer};
+use chrono::{Duration, Local};
+use serde_json::json;
+use std::ops::Drop;
+
+static DEFAULT_TOPIC_NAME: &str = "shared-resources-usage";
+/// This is the entry point for the library. It is in most cases
+/// everything you need to instrument your application.
+///
+/// The UsageAccountant needs a `Producer` and a `UsageAccumulator`.
+/// Data is stored in the accumulator and, periodically, it is
+/// flushed into the Kafka topic via a consumer.
+///
+/// Accumulating data locally is critical to reduce the performance impact
+/// of this library to a minimum and reduce the amount of Kafka messages.
+/// This means that this structure should be instantiated rarely.
+/// Possibly only once per application (or per thread).
+///
+/// Avoid creating a UsageAccountant every time some data needs to
+/// be recorded.
+
+pub struct UsageAccountant {
+    accumulator: UsageAccumulator,
+    producer: Box<dyn Producer>,
+    topic: String,
+}
+
+impl UsageAccountant {
+    /// Instantiates a UsageAccountant from a Kafka config object.
+    /// This initialization method lets the `UsageAccountant` create
+    /// the producer and own it.
+    ///
+    /// You should very rarely change topic name or granularity.
+    pub fn new(
+        producer_config: KafkaConfig,
+        topic_name: Option<&str>,
+        granularity: Option<Duration>,
+    ) -> UsageAccountant {
+        UsageAccountant::new_with_producer(
+            Box::new(KafkaProducer::new(producer_config)),
+            topic_name,
+            granularity,
+        )
+    }
+
+    /// Leaves the responsibility to provide a producer to the
+    /// client. Most of the times you should not need to use this.
+    pub fn new_with_producer(
```

Comment:
how about marking this as test only?

Reply:
I can make the DummyProducer as test only. But not this one. 
As this library is supposed to be usable in a variety of Rust system. I would want to let the callsite provide their producer implementation instantiated if they are not ok with a rdKafka ThreadedProducer

Reply:
Ok